### PR TITLE
ch4/ipc: refactor IPC and add CMA module

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
This is a temporary PR for reference. It will be rebased and possibly split into separate PRs.

[skip warnings]

TODO:
* [x] Should we enable cma by default? Many distribution default PTRACE scope to "1", which will result in `EPERM` in `process_vm_readv`.  Thus, default on will raise many support issues.
    
    EDIT: Default `MPIR_CVAR_CH4_CMA_ENABLE` to `0`
    
* [x] `--with-ch4-shmmods=posix,xpmem,cma,gpudirect` is unintuitive. I think it is better to use individual option e.g. `--with-cma`. We already use `--with-xpmem` and `--with-cuda` (and --without-` to disable).
 
  EDIT: address these in #7040

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by yfguo** [comment]
@hzhou I am rebasing the CMA code for testing. Is the IPC cleanup commits still relevant?

**Comment by hzhou** [comment]
> @hzhou I am rebasing the CMA code for testing. Is the IPC cleanup commits still relevant?

The IPC cleanup is the main purpose of this PR. You can try cherry-pick the CMA commit for your testing.

**Comment by hzhou** [comment]
test:mpich/ch4/gpu/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/xpmem

**Comment by yfguo** [comment]
test:mpich/ch4/gpu/ofi

**Comment by yfguo** [comment]
test:mpich/ch4/xpmem

**Comment by yfguo** [comment]
I pushed a few changes to fix the GPU and non-GPU build. Also making the CMA configure option the same fashion as the rest shmmods.

Should I rebased first. Will add back the white space changes from your last update.

**Comment by yfguo** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

test:mpich/custom
netmod: ch4:ofi
config: cma

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch4/gpu/ofi
test:mpich/ch4/xpmem

test:mpich/custom
netmod: ch4:ofi
config: cma

**Comment by hzhou** [comment]
tag @raffenet for review

**Comment by hzhou** [comment]
test:mpich/ch4/xpmem

test:mpich/custom
netmod: ch4:ofi
config: cma

**Comment by hzhou** [comment]
test:mpich/ch4/xpmem

test:mpich/custom
netmod: ch4:ofi
config: cma

**Comment by hzhou** [comment]
test:mpich/ch4/xpmem

test:mpich/custom
netmod: ch4:ofi
config: cma

**Comment by hzhou** [comment]
test:mpich/ch4/xpmem

**Comment by hzhou** [comment]
@raffenet This PR is ready to go.

File: src/mpid/ch4/shm/ipc/cma/cma_post.h [hunk 1]
```diff
@@ -15,13 +15,13 @@
...
-      default     : 1
+      default     : 0
       class       : none
       verbosity   : MPI_T_VERBOSITY_USER_BASIC
       scope       : MPI_T_SCOPE_ALL_EQ
       description : >-
-        To manually disable CMA set to 0. The environment variable is valid only when the CMA
-        submodule is enabled.
+        Set to 1 to manually enable CMA. It is disabled by default because the CMA requires
+        the ptrace_scope permission, which is often disabled.
```
Comment (raffenet) [inline src/mpid/ch4/shm/ipc/cma/cma_post.h:24]: is it possible to check at runtime? i should look at what ucx does...

File: src/mpid/ch4/shm/ipc/cma/cma_post.c [hunk 2]
```diff
@@ -100,19 +100,23 @@ static int copy_iovs(pid_t src_pid, MPI_Aint src_data_sz,
 
     MPI_Aint i_src = 0, i_dst = 0;
     MPI_Aint cur_offset = 0;
+    bool got_err_truncate = false;
 
+#define MAX_IOV_SEG_SIZE 0x7ffff000
```
Comment (raffenet) [inline src/mpid/ch4/shm/ipc/cma/cma_post.c:105]: is there some external source of this limit we can link to in a comment or something?

**Review by raffenet — APPROVED** [review]
Just a few comments. The total changes are a bit overwhelming, but the cleanup is long overdue for the IPC code. Hopefully the handle cache changes don't impact #7055 much.

File: src/mpid/ch4/shm/ipc/cma/cma_post.c (hunk 2 above)
Comment (hzhou) [reply src/mpid/ch4/shm/ipc/cma/cma_post.c:105]: I didn't find any. I'll add a FIXME comment.

File: src/mpid/ch4/shm/ipc/cma/cma_post.h (hunk 1 above)
Comment (raffenet) [reply src/mpid/ch4/shm/ipc/cma/cma_post.h:24]: they manually read the ptrace_scope file... https://github.com/openucx/ucx/blob/600b468427546bdfe431dc28147ae2b2916f2a9e/src/uct/sm/scopy/cma/cma_md.c#L53

Comment (hzhou) [reply src/mpid/ch4/shm/ipc/cma/cma_post.h:24]: LOL. It's too hacky to my taste. I would rather let system installer deal with this.

File: src/mpid/ch4/shm/ipc/gpu/gpu_post.c [hunk 3]
```diff
@@ -110,177 +110,328 @@
...
+
+    void *mem_addr;
+    if (dt_contig) {
+        mem_addr = MPIR_get_contig_ptr(buf, true_lb);
+    } else {
+        mem_addr = (char *) buf;
+    }
+    MPIR_GPU_query_pointer_attr(mem_addr, &ipc_attr->u.gpu.gpu_attr);
+    if (MPL_gpu_query_pointer_is_dev(mem_addr, &ipc_attr->u.gpu.gpu_attr)) {
+        /* if it's a device buffer, we can do XPMEM or CMA IPC, so set default to SKIP */
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/mpid/ch4/shm/ipc/gpu/gpu_post.c]: Fixed.

File: src/mpid/ch4/shm/ipc/cma/cma_post.h (hunk 1 above)
Comment (raffenet) [reply src/mpid/ch4/shm/ipc/cma/cma_post.h:24]: Yeah. Certainly not worth gating this PR.

**Pull request merged**